// naive-ui/lib/form/src/FormItemCol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formItemColPropKeys = exports.formItemColProps = void 0;
const vue_1 = require("vue");
const Col_1 = require("../../legacy-grid/src/Col");
const _utils_1 = require("../../_utils");
const FormItem_1 = require("./FormItem");
exports.formItemColProps = Object.assign(Object.assign({}, Col_1.colProps), FormItem_1.formItemProps);
exports.formItemColPropKeys = (0, _utils_1.keysOf)(exports.formItemColProps);
exports.default = (0, vue_1.defineComponent)({
    name: 'FormItemCol',
    props: exports.formItemColProps,
    setup() {
        const formItemInstRef = (0, vue_1.ref)(null);
        const validate = ((...args) => {
            const { value } = formItemInstRef;
            if (value) {
                return value.validate(...args);
            }
        });
        const restoreValidation = () => {
            const { value } = formItemInstRef;
            if (value) {
                return value.restoreValidation();
            }
        };
        return {
            validate,
            restoreValidation
        };
    },
    render() {
        return (0, vue_1.h)(Col_1.default, (0, _utils_1.keep)(this.$props, Col_1.colPropKeys), {
            default: () => {
                const itemProps = (0, _utils_1.keep)(this.$props, FormItem_1.formItemPropKeys);
                return (0, vue_1.h)(FormItem_1.default, Object.assign({ ref: 'formItemInstRef' }, itemProps), this.$slots);
            }
        });
    }
});
